### PR TITLE
feat: Add common hough transform components to core 

### PR DESCRIPTION
**Pull request by goblirsc**
This PR refactors the logic originally put into ACTS by @jahreda for the example hough transform algorithm.
Migrate suitable for common usage to core. 
Fix a few bugs in the process and improve execution efficiency.
Add an additional example algorithm implementing an ATLAS muon spectrometer hough transform to test the new logic. 
CC @junggjo9 @noemina @dimitra97

### CODE REVIEW

File: Core/include/Acts/Seeding/HoughTransformUtils.hpp [hunk 1]
```diff
@@ -0,0 +1,380 @@
...
+
+#include <array>
+#include <map>
+#include <optional>
+#include <set>
+#include <unordered_set>
+
+#include "HoughVectors.hpp"
+
+#pragma once
```
Comment (paulgessinger) [inline Core/include/Acts/Seeding/HoughTransformUtils.hpp:26]: I believe this needs to go all the way to the top.

File: Core/include/Acts/Seeding/HoughTransformUtils.hpp [hunk 2]
```diff
@@ -0,0 +1,380 @@
...
+
+namespace Acts {
+namespace HoughTransformUtils {
+
+/// this type is responsible for encoding the parameters of our hough space
+using coordType = double;
```
Comment (paulgessinger) [inline Core/include/Acts/Seeding/HoughTransformUtils.hpp:32]:
```suggestion
using CoordType = double;
```

File: Core/include/Acts/Seeding/HoughTransformUtils.hpp [hunk 3]
```diff
@@ -0,0 +1,380 @@
...
+
+// this type is used to encode hit counts.
+// Floating point to allow hit weights to be applied
+using yieldType = float;
```
Comment (paulgessinger) [inline Core/include/Acts/Seeding/HoughTransformUtils.hpp:36]:
```suggestion
using YieldType = float;
```

File: Core/include/Acts/Seeding/HoughTransformUtils.hpp [hunk 4]
```diff
@@ -0,0 +1,380 @@
...
+
+/// @brief this function represents a mapping of a coordinate point in detector space to a line in
+/// hough space. Given the value of the first hough coordinate, it shall return
+/// the corresponding second coordinate according to the line parametrisation.
+/// Should be implemented by the user.
+/// @tparam PointType: The class representing a point in detector space (can differ between implementations)
+template <class PointType>
+using lineParametrisation =
```
Comment (paulgessinger) [inline Core/include/Acts/Seeding/HoughTransformUtils.hpp:44]:
```suggestion
using LineParametrisation =
```

File: Core/include/Acts/Seeding/HoughTransformUtils.hpp [hunk 5]
```diff
@@ -0,0 +1,380 @@
...
+    std::function<coordType(coordType, const PointType&)>;
+
+/// @brief struct to define the ranges of the hough histogram.
+/// Used to move between parameter and bin index coordinates.
+/// Disconnected from the hough plane binning to be able to re-use
+/// a plane with a given binning for several parameter ranges
+struct houghAxisRanges {
```
Comment (paulgessinger) [inline Core/include/Acts/Seeding/HoughTransformUtils.hpp:51]:
```suggestion
struct HoughAxisRanges {
```

File: Core/include/Acts/Seeding/HoughVectors.hpp [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+#pragma once
+
+#include "Acts/Utilities/Logger.hpp"
+
+#include <sstream>
+#include <stdexcept>
+#include <vector>
+
+template <typename T>
+class vector2D {
```
Comment (paulgessinger) [inline Core/include/Acts/Seeding/HoughVectors.hpp:18]: This is a very generic name, and we have a `Vector2` already. Can we name this something that is more specific to what it's accomplishing?

File: Core/include/Acts/Seeding/HoughVectors.hpp [hunk 7]
```diff
@@ -0,0 +1,161 @@
...
+
+  void reset(const T& t) { m_data.assign(m_data.size(), t); }
+
+  T* data() { return m_data.data(); }
+
+  const T* data() const { return m_data.data(); }
+};
+
+template <typename T>
+class vector3D {
```
Comment (paulgessinger) [inline Core/include/Acts/Seeding/HoughVectors.hpp:112]: Same as for `vector2D`

File: Examples/Io/Csv/src/CsvSimHitReader.cpp [hunk 8]
```diff
@@ -60,7 +60,9 @@ ActsExamples::ProcessCode ActsExamples::CsvSimHitReader::read(
   SimHitContainer::sequence_type unordered;
   SimHitData data;
 
+  std::cout << "start to read hits " << std::endl;
```
Comment (paulgessinger) [inline Examples/Io/Csv/src/CsvSimHitReader.cpp:63]:
```suggestion
  ACTS_DEBUG("start to read hits ");
```

File: Examples/Io/Csv/src/CsvSimHitReader.cpp [hunk 9]
```diff
@@ -60,7 +60,9 @@ ActsExamples::ProcessCode ActsExamples::CsvSimHitReader::read(
...
   while (reader.read(data)) {
+    std::cout << "hi i found a sim hit" << std::endl;
```
Comment (paulgessinger) [inline Examples/Io/Csv/src/CsvSimHitReader.cpp:65]:
```suggestion
```

File: Examples/Scripts/Python/muon_hough.py
Comment (paulgessinger) [inline Examples/Scripts/Python/muon_hough.py:1]: Can we put sensible inputs for this somewhere publicly? I'd be keen to include this in the CI as soon as feasible so we can keep track of this.

**Review by paulgessinger — COMMENTED** [review]
Looks good overall. For the types that go into core, we have slightly more strict naming rules. I added a few comments on this. At the Examples level, this is much more loosely enforced, because it's not shipped.

File: Core/include/Acts/Seeding/HoughTransformUtils.hpp (hunk 1 above)
Comment (goblirsc) [reply Core/include/Acts/Seeding/HoughTransformUtils.hpp]: thanks, fixed!

File: Core/include/Acts/Seeding/HoughTransformUtils.hpp (hunk 2 above)
Comment (goblirsc) [reply Core/include/Acts/Seeding/HoughTransformUtils.hpp]: thanks, fixed!

File: Core/include/Acts/Seeding/HoughTransformUtils.hpp (hunk 4 above)
Comment (goblirsc) [reply Core/include/Acts/Seeding/HoughTransformUtils.hpp]: thanks, fixed!

File: Core/include/Acts/Seeding/HoughTransformUtils.hpp (hunk 5 above)
Comment (goblirsc) [reply Core/include/Acts/Seeding/HoughTransformUtils.hpp]: thanks, fixed!

File: Core/include/Acts/Seeding/HoughTransformUtils.hpp (hunk 3 above)
Comment (goblirsc) [reply Core/include/Acts/Seeding/HoughTransformUtils.hpp]: thanks, fixed!

File: Core/include/Acts/Seeding/HoughVectors.hpp (hunk 6 above)
Comment (goblirsc) [reply Core/include/Acts/Seeding/HoughVectors.hpp]: tried to find a nicer one!

File: Core/include/Acts/Seeding/HoughVectors.hpp (hunk 7 above)
Comment (goblirsc) [reply Core/include/Acts/Seeding/HoughVectors.hpp]: tried a nicer one

File: Examples/Io/Csv/src/CsvSimHitReader.cpp (hunk 9 above)
Comment (goblirsc) [reply Examples/Io/Csv/src/CsvSimHitReader.cpp]: done!

File: Examples/Io/Csv/src/CsvSimHitReader.cpp (hunk 8 above)
Comment (goblirsc) [reply Examples/Io/Csv/src/CsvSimHitReader.cpp]: done

File: Examples/Scripts/Python/muon_hough.py
Comment (goblirsc) [reply Examples/Scripts/Python/muon_hough.py:1]: Absolutely! What is being used in other such cases? CVMFS is probably not sufficiently general? It's no problem to store a small CSV for a few events.

**Comment by goblirsc** [comment]
Dear @paulgessinger, many thanks for the first look at this MR! One Q, do I need to do a manual squash & rebase, or will this be done later?

**Comment by paulgessinger** [comment]
This happens at the end, no need to do it manually!

**Comment by asalzburger** [comment]
uuups, I think I did a merge fu ...

File: Examples/Scripts/Python/muon_hough.py
Comment (goblirsc) [reply Examples/Scripts/Python/muon_hough.py:1]: -> now added a pair of test CSVs for one single-muon event in the format expected by the reader of the examples framework to the `Tests/Data` directory

File: Core/include/Acts/Seeding/HoughTransformUtils.hpp [hunk 10]
```diff
@@ -0,0 +1,379 @@
+// This file is part of the Acts project.
+//
+// Copyright (C) 2022 CERN for the benefit of the Acts project
+//
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+// @file HoughTransformSeeder.hpp
+// @author Max Goblirsch, based on work by Riley Xu and Jahred Adelman
```
[the remark this answers is not in the document]
Comment (goblirsc) [reply Core/include/Acts/Seeding/HoughTransformUtils.hpp:10]: removed this

File: Core/include/Acts/Seeding/HoughTransformUtils.hpp [hunk 11]
```diff
@@ -0,0 +1,379 @@
...
+
+// this type is used to encode hit counts.
+// Floating point to allow hit weights to be applied
+using YieldType = float;
+
+/// @brief this function represents a mapping of a coordinate point in detector space to a line in
+/// hough space. Given the value of the first hough coordinate, it shall return
+/// the corresponding second coordinate according to the line parametrisation.
+/// Should be implemented by the user.
+/// @tparam PointType: The class representing a point in detector space (can differ between implementations)
```
[the remark this answers is not in the document]
Comment (goblirsc) [reply Core/include/Acts/Seeding/HoughTransformUtils.hpp:41]: The CI seems to be happy, and at least VSCode correctly detects the param with the colon. 
I did not manage to locally build the doc without downloading another huge docker image, which I am not fond of. 
So tentatively replying "yeeeees... I think so? " here 😇  If we did find the doc broken after merging, I will fix it, promised!

**Review by paulgessinger — APPROVED** [review]

**Comment by acts-project-service** [comment]
## :red_circle: Athena integration test results [785654bae52583990f30e088b6f6c59776b7a703]
### :red_circle: Some tests have failed!
Please investigate the [pipeline](https://gitlab.cern.ch/acts/acts-athena-ci/-/pipelines/6774214)!

|status|job|report|
|------|---|------|
|:red_circle: | [run_unit_tests](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/35473848)||
|:red_circle: | [run_ci_tests: ../athena/Tracking/Acts/ActsConfig/test/ActsEFTrackFit.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/35472167)||
|:red_circle: | [run_ci_tests: ../athena/Tracking/Acts/ActsConfig/test/ActsBenchmarkWithSpot.sh 8 100](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/35472166)||
|:red_circle: | [run_ci_tests: ../athena/Tracking/Acts/ActsConfig/test/ActsWorkflow.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/35472165)||
|:red_circle: | [run_ci_tests: ../athena/Tracking/Acts/ActsConfig/test/ActsValidateAmbiguityResolution.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/35472164)||
|:red_circle: | [run_ci_tests: ../athena/Tracking/Acts/ActsConfig/test/ActsValidateResolvedTracks.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/35472163)||
|:red_circle: | [run_ci_tests: ../athena/Tracking/Acts/ActsConfig/test/ActsValidateTracks.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/35472162)||
|:red_circle: | [run_ci_tests: ../athena/Tracking/Acts/ActsConfig/test/ActsValidateActsCoreSpacePoints.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/35472161)||
|:red_circle: | [run_ci_tests: ../athena/Tracking/Acts/ActsConfig/test/ActsValidateActsSpacePoints.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/35472160)||
|:red_circle: | [run_ci_tests: ../athena/Tracking/Acts/ActsConfig/test/ActsValidateSeeds.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/35472159)||
|:red_circle: | [run_ci_tests: ../athena/Tracking/Acts/ActsConfig/test/ActsValidateOrthogonalSeeds.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/35472158)||
|:red_circle: | [run_ci_tests: ../athena/Tracking/Acts/ActsConfig/test/ActsValidateClusters.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/35472157)||
|:red_circle: | [run_ci_tests: ../athena/Tracking/Acts/ActsConfig/test/ActsPersistifyEDM.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/35472156)||
|:red_circle: | [run_ci_tests: ../athena/Tracking/Acts/ActsConfig/test/ActsGSFRefitting.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/35472155)||
|:red_circle: | [run_ci_tests: ../athena/Tracking/Acts/ActsConfig/test/ActsKfRefitting.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/35472154)||
|:red_circle: | [run_ci_tests: python3 ../athena/Tracking/Acts/ActsGeometry/test/ActsExtrapolationAlgTest.py](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/35472153)||
|:red_circle: | [run_ci_tests: python3 ../athena/Tracking/Acts/ActsGeometry/test/ActsITkTest.py](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/35472152)||
|:red_circle: | [run_workflow_tests_run4_mc](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/35472151)||
|:red_circle: | [run_workflow_tests_run2_mc](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/35472150)||
|:red_circle: | [run_workflow_tests_run2_data](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/35472149)||
|:red_circle: | [run_workflow_tests_run3_mc](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/35472148)||
|:red_circle: | [run_workflow_tests_run3_data](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/35472147)||
|:red_circle: | [run_art_test: test_data18_13TeV_1000evt](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/35472146)||
|:red_circle: | [run_art_test: test_ttbarPU40_reco](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/35472145)||

**Comment by paulgessinger** [comment]
Athena failure is unrelated to changes, ATLAS / CERN is having infrastructure issues.

**Pull request merged**